compiler/symbol.go

```diff
@@ -375,19 +375,28 @@ func (c *compilerContext) checkWasmImport(f *ssa.Function, pragma string) {
 
 // Check whether the type maps directly to a WebAssembly type, according to:
 // https://github.com/golang/go/issues/59149
+//
+// Update: this reflects the relaxed type restrictions proposed here:
+// https://github.com/golang/go/issues/66984
 func isValidWasmType(typ types.Type, isReturn bool) bool {
 	switch typ := typ.Underlying().(type) {
 	case *types.Basic:
 		switch typ.Kind() {
-		case types.Int32, types.Uint32, types.Int64, types.Uint64:
+		case types.Bool:
+			return true
+		case types.Int8, types.Uint8, types.Int16, types.Uint16, types.Int32, types.Uint32, types.Int64, types.Uint64:
 			return true
 		case types.Float32, types.Float64:
 			return true
-		case types.UnsafePointer:
-			if !isReturn {
-				return true
-			}
+		case types.Uintptr, types.UnsafePointer:
+			return true
+		case types.String:
+			return true
 		}
+	case *types.Struct:
+		return true
```

Comment:
This is actually not part of https://github.com/golang/go/issues/66984 anymore (though it is when it's a pointer-to-struct and pointer-to-array, this probably needs some more logic):

> ### Supporting `struct` and `[...]T` by value
> 
> A previous version of this proposal included support for passing `struct` and `[...]T` types by value by expanding each field recursively into call parameters. This was removed in favor of a simpler initial implementation but could be readded if users require it.

Reply:
Also, now I think about it, we should _definitely_ be stricter here and also check every field in the struct. Otherwise `func(x *struct{field [2]int})` would be legal which it shouldn't be.